### PR TITLE
persistent peers (Refs 13)

### PR DESCRIPTION
**Pull request by melekes**
Tests for MConnection plus [persistent peers](https://github.com/tendermint/go-p2p/pull/18/commits/b33f7b7d9ecfd7f19b20f817679930014e5dfa12).

### CODE REVIEW

File: peer.go [hunk 1]
```diff
@@ -4,134 +4,291 @@ import (
...
+}
+
+// MakePersistent marks the peer as persistent.
+func (p *Peer) MakePersistent() {
+	p.persistent = true
+}
+
+// IsPersistent returns true if the peer is persitent, false otherwise.
+func (p *Peer) IsPersistent() bool {
+	return p.persistent && !p.outbound
```
Comment (ebuchman) [inline peer.go]: not outbound?

File: peer.go [hunk 2]
```diff
@@ -4,134 +4,291 @@ import (
...
+// IsOutbound returns true if the connection is outbound, false otherwise.
 func (p *Peer) IsOutbound() bool {
 	return p.outbound
 }
 
+// Send msg to the channel identified by chID byte. Returns false if the send
+// queue is full after timeout, specified by MConnection.
 func (p *Peer) Send(chID byte, msg interface{}) bool {
 	if !p.IsRunning() {
+		// FIXME shouldn't we panic here?
```
Comment (ebuchman) [inline peer.go]: Don't think so. We do seem to remove the peer before we call Stop, but see sw.Broadcast, where we fetch the list of peers and loop over them - while we're looping, one peer may be removed and stopped, in which case the Send will fail.

File: switch.go [hunk 3]
```diff
@@ -340,22 +307,22 @@ func (sw *Switch) dialSeed(addr *NetAddress) {
...
+		log.Info("Failed dialing peer", "address", addr, "error", err)
+		peer.CloseConn()
 		return nil, err
 	}
-	if sw.config.GetBool(configFuzzEnable) {
-		conn = FuzzConn(sw.config, conn)
-	}
-	peer, err := sw.AddPeerWithConnection(conn, true)
+	err = sw.AddPeer(peer)
+	sw.dialing.Delete(addr.IP.String())
```
Comment (ebuchman) [inline switch.go]: Shouldn't this come right after `newPeer` ? Otherwise we might return an error and be stuck thinking we're still dialing?

Comment (ebuchman) [inline switch.go]: Actually, maybe we should pass the `dialing Cmap` to `func dial` ... `Redial` doesn't seem to set the peer in `dialing`

File: peer.go (hunk 1 above)
Comment (melekes) [inline peer.go]: https://github.com/tendermint/go-p2p/blob/29da4e820cf67cc4f94ecb8c93b69a113322843d/peer.go#L21

Comment (melekes) [inline peer.go]: I will rewrite it to panic instead on `MakePersistent`

File: peer.go (hunk 2 above)
Comment (melekes) [inline peer.go]: Ok, I will add a comment if you don't mind

File: switch.go (hunk 3 above)
Comment (melekes) [inline switch.go]: Yes, good catch!

File: peer.go (hunk 1 above)
Comment (ebuchman) [inline peer.go]: Ok, but, doesn't persistent require that the peer is outbound?

File: peer.go (hunk 2 above)
Comment (ebuchman) [inline peer.go]: Please do

File: peer.go (hunk 1 above)
Comment (melekes) [inline peer.go]: oh boy.. you are right

File: switch_test.go [hunk 4]
```diff
@@ -205,8 +214,14 @@ func TestConnPubKeyFilter(t *testing.T) {
 	})
 
 	// connect to good peer
-	go s1.AddPeerWithConnection(c1, false) // AddPeer is blocking, requires handshake.
-	go s2.AddPeerWithConnection(c2, true)
+	go func() {
+		peer, _ := newPeerFromExistingConn(c1, false, s1.reactorsByCh, s1.chDescs, s1.StopPeerForError, s1.config, s1.nodePrivKey)
```
Comment (ebuchman) [inline switch_test.go]: we should have AddPeerWithConnection as a helper function that just calls newPeerFromExistingConn and AddPeer. Then we don't have to get so messy here :)

Comment (melekes) [inline switch_test.go]: Agree

Comment (ebuchman) [inline switch_test.go]: alternatively, or additionally, should we make these `newPeer` funcs methods on the switch? They take a lot of switch fields...

File: switch.go [hunk 5]
```diff
@@ -313,13 +313,13 @@ func (sw *Switch) DialPeerWithAddress(addr *NetAddress, persistent bool) (*Peer,
 	if persistent {
 		peer.MakePersistent()
 	}
+	sw.dialing.Delete(addr.IP.String())
```
Comment (ebuchman) [inline switch.go]: what about in Redial ...  we don't update the dialing map there

File: switch_test.go (hunk 4 above)
Comment (melekes) [inline switch_test.go]: I was planning to refactor them... maybe after we release?

File: switch.go (hunk 5 above)
Comment (melekes) [inline switch.go]: fixed in https://github.com/tendermint/go-p2p/pull/18/commits/837b0ed702805815e5cc9ac04599292a222d73db

File: switch_test.go (hunk 4 above)
Comment (ebuchman) [inline switch_test.go]: Ok cool

File: peer.go [hunk 6]
```diff
@@ -4,134 +4,260 @@ import (
...
+func (p *Peer) IsPersistent() bool {
+	return p.persistent
+}
+
+// HandshakeTimeout performs a handshake between a given node and the peer.
 // NOTE: blocking
-// Before creating a peer with newPeer(), perform a handshake on connection.
-func peerHandshake(conn net.Conn, ourNodeInfo *NodeInfo) (*NodeInfo, error) {
+func (p *Peer) HandshakeTimeout(ourNodeInfo *NodeInfo, timeout time.Duration) error {
+	// Set deadline for handshake so we don't block forever on conn.ReadFull
```
Comment (ebuchman) [inline peer.go]: Ok ... now I'm not sure we need this here. Would ReadBinary block forever or just fail? Whether or not we need it here, we definetely need one before we authenticate/encrypt the connection. I can add that

**Pull request merged**